pkg/objstore/objstore.go

```diff
@@ -254,34 +330,55 @@ func DownloadFile(ctx context.Context, logger log.Logger, bkt BucketReader, src,
 }
 
 // DownloadDir downloads all object found in the directory into the local directory.
-func DownloadDir(ctx context.Context, logger log.Logger, bkt BucketReader, originalSrc, src, dst string, ignoredPaths ...string) error {
+func DownloadDir(ctx context.Context, logger log.Logger, bkt BucketReader, originalSrc, src, dst string, options ...DownloadDirOption) error {
 	if err := os.MkdirAll(dst, 0750); err != nil {
 		return errors.Wrap(err, "create dir")
 	}
+	opts := applyDownloadDirOptions(options...)
+
+	g, ctx := errgroup.WithContext(ctx)
+	guard := make(chan struct{}, opts.concurrency)
 
 	var downloadedFiles []string
-	if err := bkt.Iter(ctx, src, func(name string) error {
-		dst := filepath.Join(dst, filepath.Base(name))
-		if strings.HasSuffix(name, DirDelim) {
-			if err := DownloadDir(ctx, logger, bkt, originalSrc, name, dst, ignoredPaths...); err != nil {
+	var m sync.Mutex
+
+	err := bkt.Iter(ctx, src, func(name string) error {
+		guard <- struct{}{}
+		g.Go(func() error {
+			defer func() { <-guard }()
+			dst := filepath.Join(dst, filepath.Base(name))
+			if strings.HasSuffix(name, DirDelim) {
+				if err := DownloadDir(ctx, logger, bkt, originalSrc, name, dst, options...); err != nil {
+					return err
+				}
+				m.Lock()
+				defer m.Unlock()
```

Comment:
This lock is just to guard `downloadedFiles`, no? I would rather we minimize the critical section rather than use `defer`.

Reply:
We will return at line 357 and the next defer will return at line 372. I think this should be fine, but also okay to change to make it clearer.

Reply:
Yeah.. i can make more clear.